src/mlpack/core/optimizers/adam/adam_impl.hpp

```diff
@@ -113,19 +119,44 @@ double Adam<DecomposableFunctionType>::Optimize(arma::mat& iterate)
     m *= beta1;
     m += (1 - beta1) * gradient;
 
-    v *= beta2;
-    v += (1 - beta2) * (gradient % gradient);
+    if (adaMax)
+    {
+      u *= beta2;
+      arma::mat absGradient = arma::abs(gradient);
+      // Update the exponentially weighted infinity norm.
+      for (size_t row = 0; row != iterate.n_rows; ++row)
```

Comment:
There are some side effects like branch predictions, memory usage, etc. that has an influence on the runtime of the code that should theoretically be faster.

But it sounds like you had fun to explore another approach, and if it turns out to be faster even better.